sw/host/tests/chip/usb/usbdev_suspend.rs

```diff
@@ -0,0 +1,357 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// This harness controls a number of suspend-sleep-resume USBDEV tests:
+// - usbdev_suspend_resume_test
+// - usbdev_sleep_resume_test
+// - usbdev_sleep_reset_test
+// - usbdev_sleep_disconnect_test
+// - usbdev_sleep_resume_test
+// - usbdev_sleep_reset_test
+// - usbdev_sleep_disconnect_test
+//
+// The above are sub-sections of the following full sequence test:
+// - usbdev_suspend_full_test
+
+use anyhow::{anyhow, bail, ensure, Context, Result};
+use clap::{Parser, ValueEnum};
+use std::time::Duration;
+
+use opentitanlib::app::TransportWrapper;
+use opentitanlib::execute_test;
+use opentitanlib::io::uart::Uart;
+use opentitanlib::test_utils::init::InitializeTest;
+use opentitanlib::uart::console::UartConsole;
+
+use usb::{UsbHub, UsbHubOp, UsbOpts};
+
+#[derive(Debug, Parser)]
+struct Opts {
+    #[command(flatten)]
+    init: InitializeTest,
+
+    // Console/USB timeout; test harness must allow time for device configuration by
+    // the host, as well as iterating through a number of test phases.
+    #[arg(long, value_parser = humantime::parse_duration, default_value = "60s")]
+    timeout: Duration,
+
+    /// USB options.
+    #[command(flatten)]
+    usb: UsbOpts,
+
+    // Iteration count.
+    #[arg(long, default_value = "1")]
+    num_iters: u32,
+
+    // Test phases.
+    #[arg(long)]
+    init_phase: SuspendPhase,
+    #[arg(long)]
+    fin_phase: SuspendPhase,
+}
+
+// Test phases
+// - the tests controlled by this sequence presently run a contiguous sequence of phases
+//   within this full sequence.
+#[derive(Clone, Debug, PartialEq, Eq, ValueEnum)]
+enum SuspendPhase {
+    // Resume Signaling stimulus when suspended but not in a sleep state.
+    Suspend,
+    // Resume Signaling when the device is in Normal Sleep
+    //  (clocks stopped but USBDEV not powered down).
+    SleepResume,
+    // Awaken from Normal Sleep in responses to a Bus Reset.
+    SleepReset,
+    // ... to a VBUS removal (disconnection).
+    SleepDisconnect,
+    // As for the above tests phases, but entering Deep Sleep rather in which power is removed also.
+    DeepResume,
+    DeepReset,
+    DeepDisconnect,
+    // Final test state; device disconnects and test completes.
+    Shutdown,
+}
+
+impl ToString for SuspendPhase {
```

Comment:
Should use the crate strum to implement this for you
https://github.com/lowRISC/opentitan/blob/405b21e625e7481a0d94b0c8077a1c3f1d1373f1/sw/host/opentitanlib/src/dif/lc_ctrl.rs#L316